### ISSUE TITLE
JAX-RS subresource issue in Quarkus native build

### ISSUE THREAD

**Issue by DeMol-EE** [quarkusio/quarkus#37148]
### Describe the bug

General architecture:

```java
@Path("foo")
@RequestScoped
public class FooResource {

  @Path("{id}")
  public BarResource bar(@Context ResourceContext rc) {
        return rc.initResource(CDI.current()
                .select(BarResource.class)
                .get());
  }
}
```

```java
@Dependent
@Unremovable
@RegisterForReflection
public class BarResource {

  @PathParam("id")
  String id;

  // some CDI @Inject dependencies omitted

  @GET
  public boolean get(@QueryParam("bar") Integer bar) {
    return true;
  }
}
```

When executing an HTTP GET for /foo/some?bar=5, we get the following exception:

```
Encountered an issue: org.jboss.resteasy.spi.InternalServerErrorException: RESTEASY003815: Subresource for target class has no qualifying annotations.: ...BarResource.class...
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:157)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:131)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:33)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:429)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:240)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:154)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:321)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:157)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:229)
	at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:82)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:147)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$1.run(VertxRequestHandler.java:93)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:582)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base@21.0.1/java.lang.Thread.runWith(Thread.java:1596)
	at java.base@21.0.1/java.lang.Thread.run(Thread.java:1583)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:832)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.posix.thread.PosixPlatformThreads.pthreadStartRoutine(PosixPlatformThreads.java:211)
```

### Expected behavior

No exception, or detection of the issue at build time.

### Actual behavior

Exception when trying to call endpoint specified in subresource

### How to Reproduce?

See [reproducer](https://github.com/DeMol-EE/quarkus-reproducer-37148)

### Output of `uname -a` or `ver`

runtime provided by quarkus-micro-image:2.0

### Output of `java -version`

jdk 21 used in the mandrel build image

### Mandrel or GraalVM version (if different from Java)

built with quay.io/quarkus/ubi-quarkus-mandrel-builder-image:jdk-21, released with quay.io/quarkus/quarkus-micro-image:2.0

### Quarkus version or git rev

3.5.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

mvn 3.9.5, via mvnw 3.2.0

### Additional information

* JAXRS extension: quarkus-resteasy-jackson
* We added `@RegisterForReflection` after reading [this](https://stackoverflow.com/questions/57820428/jax-rs-subresource-issue-in-quarkus), but to no avail
* Getting subresources and CDI to work never went smoothly, it took some trial-and-error but the shenanigans with `ResourceContext` work.
* this is a minimal Quarkus app with strictly quarkus extensions as dependencies

**Comment by DeMol-EE** [comment]
Another message I’m getting:

...(executor-thread-1) Encountered an issue: java.lang.RuntimeException: RESTEASY003875: Unable to find a constructor that takes a String param or a valueOf() or fromString() method for jakarta.ws.rs.QueryParam("someId") on public boolean ...FooResource.fooMethod(boolean,java.lang.Integer) for basetype: java.lang.Integer

**Comment by galderz** [comment]
@DeMol-EE Do you have a reproducer?

**Comment by DeMol-EE** [comment]
I’ve created one: https://github.com/DeMol-EE/quarkus-reproducer-37148

Run with `mvn quarkus:dev` and do `curl "http://localhost:8080/hello/foo?bar=5"`, expect to receive the response "Hello foo 5".

Package with `docker build -t reproducer-37148 .`, run with `docker run --rm -p 8080:8080 reproducer-37148` and do the same curl command as above to observe the error.

**Comment by DeMol-EE** [comment]
I’ve noticed something new. It appears that on each HTTP request, a new instance of BarResource (the subresource) is created (as expected as it’s a @Dependent bean), but it is not being cleaned up after the response is sent. I’ve added the following to both resources:

```java
    @PostConstruct
    public void hi() {
        System.out.println(">" + this.getClass().getSimpleName());
    }

    @PreDestroy
    public void bye() {
        System.out.println("<" + this.getClass().getSimpleName());
    }
```

and on each request, the following is logged:

```
>FooResource_Subclass
>BarResource_Subclass
<FooResource_Subclass
```

implying that BarResource is being retained somewhere. On quitting quarkus (stopping dev mode by hitting q), we can see the @PreDestroy hook of the subresource is being called and the instances are being released.

Perhaps the leak is somehow resulting in an issue in the native compilation but this is a wild guess (I really have very little knowledge in this area).

We found this while investigating a memory leak in our application. [This thread](https://github.com/quarkusio/quarkus/issues/25083) has been inspirational, but I don’t understand what is retaining the instances in our case because nothing is @ApplicationScoped. I’m fairly certain it has to do with the programmatic CDI lookup, but I did not expect this. I would have expected that the injected bean follows the lifecycle of the resource. I’d love to hear @mkouba’s thoughts on this. I’ve updated the reproducer.

Note that we have been able to get around the issue by doing any one of the following:
* use `@Inject` to inject an instance of the sub resource in the parent resource resolves the issue (the dependent sub resource has its lifecycle bound properly)
* rescope the sub resource as `@RequestScoped` also resolves the issue
* flatted the sub resource, inlining it into the parent

There is still a memory leak in our application and I’m starting to lean towards making everything application scoped by default, instead of @Dependent... but that’s my concern.

**Comment by DeMol-EE** [comment]
Observation: moving from RESTEasy classic to RESTEasy reactive fixes the native build!

**Comment by DeMol-EE** [comment]
bump @galderz

**Comment by michalvavrik** [comment]
> Observation: moving from RESTEasy classic to RESTEasy reactive fixes the native build! However, the @PathParam in the sub resource is not set (it’s always null, even in jvm mode), though that is to be expected according to the spec, as CDI act as creator in our case:

cc @geoand

**Comment by geoand** [comment]
Please open a new issue for the RESTEasy Reactive problem

**Comment by geoand** [comment]
> It is a little surprising that CDI does not set the fields, but perhaps we'll get that in the (near) future, with JAXRS 4.
The retention of the dependent sub resource is still there though...

Jakarta REST until recently do not have any OOTB integration with CDI, so this is expected (RESTEasy Reactive however does have this integration built in from day 1).

**Comment by DeMol-EE** [comment]
@geoand I see, but I'd like to point out that neither implementation sets fields annotated with `@PathParam` in sub-resources. With RESTEasy classic, injecting `ResourceContext` and calling `initResource` (as shown in the original post) did the trick, but this no longer works for RESTEasy Reactive. Am I to understand that with reactive `@PathParam` fields _should_ be injected? And is that the problem you are referring to when asking to create a new ticket? Or are you referring to the seeming memory leak when doing a programmatic lookup of `@Dependent` sub-resources?

**Comment by geoand** [comment]
> Am I to understand that with reactive @PathParam fields should be injected? And is that the problem you are referring to when asking to create a new ticket

Right this.

**Comment by tanadeau** [comment]
@DeMol-EE: I'm seeing the same problem with `@PathParam` injection for subresources. I couldn't find a new ticket. Did you create one? If not, I was going to alongside a reproducer next week.

**Comment by DeMol-EE** [comment]
@tanadeau I have not made a new ticket yet, so go ahead. For what it’s worth, we settled for a workaround which is to inject the path params in the root resource, and passing them to the sub resource instance via setters. We also ended up making sub resources `@RequestScoped`, so we can use them as short-lived, private (stateful) instances.

**Comment by zakkak** [comment]
@DeMol-EE can you still reproduce the original issue? It seems to work as expected with Quarkus 3.9.1

**Comment by tanadeau** [comment]
@zakkak I’m still seeing the issue on 3.9.1.

**Comment by tanadeau** [comment]
@DeMol-EE @zakkak: That’s what I’m seeing as well.

**Comment by DeMol-EE** [comment]
Will do. I’ll create 2 in fact: one for path param not being injected in sub resources, and one for dependent sub resources being retained forever (which might not be a bug but I can’t explain why they are being retained).

**Comment by tanadeau** [comment]
@DeMol-EE: I had some time, and I went ahead and created the ticket about the path param injection issue.

**Comment by DeMol-EE** [comment]
I tried to reproduce this in Quarkus 3.32.3 and now get this build-time exception:

```
[error]: Build step io.quarkus.arc.deployment.ArcProcessor#registerBeans threw an exception: jakarta.enterprise.inject.spi.DeploymentException: Resource classes that use field injection for REST parameters can only be @RequestScoped. Offending class is org.acme.BarResource
```

It is now impossible to create the original issue, which is a good thing, because when following the advice and making the sub resource `@RequestScoped`, the application no longer suffers from a memory leak. I tried a few variations to see if I could still somehow trigger a memory leak but whatever I did, everything worked as expected. Even with `@Dependent` and sub resources.

Keep up the good work!
